minimalkv/fsspecstore.py

```diff
@@ -0,0 +1,207 @@
+        return self._file.seekable()
+
+    def readable(self):  # noqa
+        return self._file.readable()
+
+    def close(self) -> None:
+        """Close the file."""
+        self._file.close()
+
+    def closed(self) -> bool:
+        """Whether the file is closed."""
+        return self._file.closed
+
+
+class FSSpecStore(KeyValueStore):
+    """A KeyValueStore that uses an fsspec AbstractFileSystem to store the key-value pairs."""
+
+    def __init__(self, fs: AbstractFileSystem, prefix="", mkdir_prefix=True):
+        """
+        Initialize an FSSpecStore.
+
+        Parameters
+        ----------
+        fs: AbstractFileSystem
+            The fsspec filesystem to use.
+        prefix: str, optional
+            The prefix to use on the FSSpecStore when storing keys.
+        mkdir_prefix : Boolean
+            If True, the prefix will be created if it does not exist.
+            Analogous to the create_if_missing parameter in AzureBlockBlobStore or GoogleCloudStore.
+        """
+        self.fs = fs
+        self.prefix = prefix
+
+        if mkdir_prefix:
+            self.fs.mkdir(self.prefix)
+
+    @lazy_property
+    def _prefix_exists(self):
+        # Check if prefix exists
+        try:
+            self.fs.info(self.prefix)
+        except (FileNotFoundError, IOError):
+            return False
+        return True
+
+    def iter_keys(self, prefix: str = "") -> Iterator[str]:
+        """Iterate over all keys in the store starting with prefix.
+
+        Parameters
+        ----------
+        prefix: str, optional, default = ''
+            Only iterate over keys starting with prefix. Iterate over all keys if empty.
+
+        Raises
+        ------
+        IOError
+            If there was an error accessing the store.
+        """
+        # List files
+        all_files_and_dirs = self.fs.find(f"{self.prefix}", prefix=escape(prefix))
+
+        # When no matches are found, the Azure FileSystem returns the container,
+        # which is not desired.
+        if len(all_files_and_dirs) == 1 and all_files_and_dirs[0] in self.prefix:
+            return iter([])
+        return map(
+            lambda k: unescape(k.replace(f"{self.prefix}", "")), all_files_and_dirs
+        )
+
+    def _delete(self, key: str):
+        try:
+            self.fs.rm_file(f"{self.prefix}{escape(key)}")
+        except FileNotFoundError:
+            pass
+
+    def _open(self, key: str) -> IO:
+        if not self._prefix_exists:
+            raise NotFound("Bucket does not exist.")
+        try:
+            return self.fs.open(f"{self.prefix}{escape(key)}")
+        except FileNotFoundError:
+            raise KeyError(key)
+
+    def _put_file(self, key: str, file: IO) -> str:
+        # fs.put_file only supports a path as a parameter, not a file
+        # Open file at key in writable mode
+        with self.fs.open(f"{self.prefix}{escape(key)}", "wb") as f:
+            copyfileobj(file, f)
+            return key
```

Comment:
I think you should be able to use [`pipe`](https://github.com/fsspec/filesystem_spec/blob/3969aaf7378198b92df8dee99cd464d272a0d0f4/fsspec/spec.py#L727-L747)

Reply:
Thx!